Sources/TensorFlow/Operators/Basic.swift

```diff
@@ -390,44 +390,92 @@ public extension Tensor {
+            let posAxis = axis < 0 ? axis + rank : axis
+
+            // TODO: precondition(posAxis >= 0 && posAxis < rank, "'axis' is out of range.")
+            // TODO: precondition(batchDims <= posAxis,
+            //                    "'batchDims' must be less than or equal to 'axis'.")
+
+            // Move self[axis] up to self[batchDims].
+            let permutation = Tensor<Int32>(concatenating: [
+                Tensor<Int32>(rangeFrom: 0, to: Int32(batchDims), stride: 1),
+                Tensor<Int32>(Int32(axis)).rankLifted(),
+                Tensor<Int32>(rangeFrom: Int32(batchDims), to: Int32(posAxis), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(axis) + 1, to: Int32(rank), stride: 1)])
+            let tensor = transposed(withPermutations: permutation)
+            let result = tensor.batchGathering(
+                atIndices: indices,
+                alongAxis: batchDims,
+                batchDimCount: batchDims)
+
+            // Move the result dimensions corresponding to self[batchDims ..< axis] to just before
+            // the dimensions corresponding to indices[batchDims ...].
+            let start = indices.rank + posAxis - batchDims
+            let resultPermutation = Tensor<Int32>(concatenating: [
+                Tensor<Int32>(rangeFrom: 0, to: Int32(batchDims), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(indices.rank), to: Int32(start), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(batchDims), to: Int32(indices.rank), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(start), to: Int32(result.rank), stride: 1)])
+            return result.transposed(withPermutations: resultPermutation)
+        }
+
+        let batchIndices = withoutDerivative(at: { () -> Tensor<Index> in
+            var batchIndices = indices
+            var accumulated = Tensor<Index>(ones: [])
+            for d in (1...batchDims).reversed() {
+                accumulated *= Tensor<Index>(shapeTensor[d])
+                let dValue = shapeTensor[d - 1]
+                let dIndices = Tensor<Index>(
+                    rangeFrom: Tensor<Index>(zeros: []),
+                    to: Tensor<Index>(dValue),
+                    stride: Tensor<Index>(ones: [])
+                ) * accumulated
+                let dShape = Tensor<Int32>(concatenating: [
+                    Tensor<Int32>([Int32](repeating: 1, count: d - 1)),
+                    dValue.rankLifted(),
+                    Tensor<Int32>([Int32](repeating: 1, count: indices.rank - 1))])
+                batchIndices += dIndices.reshaped(toShape: dShape)
+            }
+            return batchIndices
+        }())
+
         let flatIndices = batchIndices.flattened()
-        let outerShape = withoutDerivative(at: shapeTensor) { $0[2...] }
-        let innerShape = withoutDerivative(at: shapeTensor) { $0[..<2] }.product(squeezingAxes: [0])
+        let outerShape = shapeTensor[Int(batchDims + 1)...]
```

Comment:
`batchDimensionCount` is already `Int`.
```suggestion
        let outerShape = shapeTensor[(batchDimensionCount + 1)...]
```